### ISSUE TITLE
issue/request -Angrylion RDP into GLideN64...

### ISSUE THREAD

**Issue by fallaha56** [gonetz/GLideN64#824]
hey guys merry xmas (almost) to you all

just wondering...with various issues in the renderer being worked and given the high+ requirements to run a full software RDP is there any scope to combine Angrylion's RDP into the plugin?

perhaps with GLideN64 doing the VI filter work in OpenGL?

would this make for a single perfect emulation N64 plugin?

**Comment by theboy181** [comment]
I think that a software plugin that is optimized would run quite nice. I hear that AIO (legend of Dragoon) has done some amazing speed optimization using SSE4, and AL. I think that AVX might be the teeter totter tipper that AL's software RDP needs.

**Comment by fallaha56** [comment]
not going to disagree with that!

but surely if MAME can multi-thread its 3Dfx games then surely someone can do the same for the RDP?

**Comment by theboy181** [comment]
I have never played with the 3dfx stuff in mame.. What do you recommend? I would like to test that out. Maybe AIO could take advantage of this type of enhancement

**Comment by fallaha56** [comment]
any of the newer Atari games -San Fran Rush, Gauntlet Leg

of note option is off by default in MAME options...

-i think it was Ville Linde who did the coding -he called it software SLI
http://vlinde.mameworld.info/

**Comment by fallaha56** [comment]
take that back, it was Haze!
http://mamedev.emulab.it/haze/

**Comment by theboy181** [comment]
Looking on youtube, most users that are running these games are on 4ghz computers.. I think that AIO's work on single CPU is far better..

**Comment by theboy181** [comment]
add me to skype if you have it...theboy181

**Comment by fallaha56** [comment]
i'm on both -4Ghz 6-core 980x and MAME uses all 6 cores heavily...

but my Surface4Pro has much less gigahurtz...

**Comment by purplemarshmallow** [comment]
This is something I thought about before. Can angrylion's plugin be used for software framebuffer emulation in a hardware renderer? I think it should be possible. Depth and auxiliary buffers can be rendered in software and main color buffers are rendered in hardware. Jabo did that in his plugin with his software renderer.

The alternative is to render everything on the GPU and copy over the necessary data. That's how GLideN64 currently works. Both methods have their advantages and disadvantages. A full hardware renderer is possibly faster.

**Comment by fallaha56** [comment]
hi @purplemarshmallow merry xmas, great minds think alike ;)

if i understand things right (and i might not!):
-the angrylion RDP is pretty much 100% accurate but CPU heavy
-GLideN64 still needs hacks+++
-once the image reaches the VI filters, it's one way traffic?
-prior to this there's lots of 2-way traffic between CPU and the RDP?

just to keep things simple and not worry about 2-way traffic why not just plug the angrylion RDP code into the GLideN64 VI? this would be both accurate and pretty fast no?

**Comment by LegendOfDragoon** [comment]
>this would be both accurate and pretty fast no?

It would not be fast.

**Comment by fallaha56** [comment]
hi @LegendOfDragoon, @theboy181 tells me you're a man of much knowledge of such things ;) 

am guessing you may know my next question already...

-i get depressingly close to 60fps on both my Surface and desktop rig in full software BUT only with VI filters off

-that's using the old plugins from the clearly talented (but the equally as difficult to talk to) Hatcat

-so...if the RDP is optimised/SSE4/AVX/multithreaded and the VI offloaded to OpenGL would we not start hitting the magical 60fps?

(plus integrate it and it could help develop the full hardware side of GLideN64?)

**Comment by LegendOfDragoon** [comment]
It depends on the game you're playing, as well as the resolution (for games that support hi-res). It could get a lot faster, but at the same time, some games are just too demanding. If you're interested in specific games, I could give you a better estimate.

**Comment by fallaha56** [comment]
sure, i appreciate it won't work well for every game but i'm still drawn to the idea of having a built-in reference renderer/a single GFX plugin for PJ64

which optimisations do you think hold back PJ64 the most/which would benefit?

(in case it's of interest i've tested Mario64, F-Zero, Zelda:OOT, MarioTennis and MK64 but with the exception of MTennis they're all that tantalising 55-60fps...)

**Comment by LegendOfDragoon** [comment]
SM64, OOT, MK64 should run full speed on average hardware, after optimizing. MarioTennis will take some work (certain parts of the game are more demanding) and F-ZERO will take a lot of work to get running on mid-low end machines (if even possible). I imagine high end machines would run F-ZERO and Mario Tennis fine though.

I'd say the order for most demanding (from highest to lowest) for most games is RDP, RSP, VI, and then CPU (assuming you're using a cpu recompiler). I think using 64-bit could make a significant difference.

**Comment by fallaha56** [comment]
interesting

64-bit seems to be close, recompiler the outstanding issue?

also, if the RDP is a big bottleneck then won't multi-threading help? MAME finally made the R5000/3Dfx games playable this way (and never thought i'd be complimenting MAME on speed!)

**Comment by fallaha56** [comment]
also need to take back what i said about F-Zero must have been on the Kool Aid at the time (or running GLideN64), it's slow as molasses in software lol

**Comment by LegendOfDragoon** [comment]
>recompiler the outstanding issue?

Recompiler makes a huge difference in speed.

>also, if the RDP is a big bottleneck then won't multi-threading help? MAME finally made the R5000/3Dfx games playable this way (and never thought i'd be complimenting MAME on speed!)

Short answer is, yes it can help. But I think that should literally be the last thing done, if at all. Focus should go towards optimizing the code as much as possible, before using multi-threading. There's just too many potential optimizations. It's just much more efficient right now to optimize the slow code, than it is to spend time working on multi threading.

**Comment by fallaha56** [comment]
thanks, appreciate the insights, 3am here tho so off to bed! happy/keen to be involved in whatever way i can. merry xmas

**Comment by gonetz** [comment]
My two kopecks: if you want to combine soft and hw renders, they both must support the same mode, HLE or LLE.
Angrylion's plugin works only in LLE. GLideN64 is not very good in LLE.

Merry Christmas!

**Comment by theboy181** [comment]
Merry Xmas @gonetz I hope SANTA is good to you this year

**Comment by gonetz** [comment]
Thanks, but I'm from Russia. Santa has no power there. Also, Russian Orthodox Church uses Julian calendar, so Christmas in Russia will be officially celebrated January 7.
Since Russia supports all main religions, many russians celebrate xmas twice, today and two weeks after. The same is with New Year holidays. "Old New Year" is unofficially celebrated December 31 night by Julian calendar, that is January 13.

**Comment by fallaha56** [comment]
Merry No.1 Christmas @gonetz, my family is Orthodox as well, two for the price of one ;)

If I'm frank, one of the biggest reasons I suggested this was that I believe fragmented work is hurting N64 emulation badly, this could help solve that problem and possibly even serve as a catalyst for another round of donations if you wanted?

Plus, to play devils advocate, does it matter if the soft RDP is LLE only if the RSP recompiler covers both?

A single unified GFX plugin that works with virtually all games? That would b something special...

**Comment by gonetz** [comment]
LLE and HLE display lists are absolutely different.  RSP plugin returns only one of them. I think it is possible to run two threads,  HLE one with hardware plugin and LLE with soft render, then sync them. However we need special core or RSP plugin for that.

**Comment by theboy181** [comment]
Zilmar just holds too much special powers over this. ;).

**Comment by purplemarshmallow** [comment]
I think it would be better to give the gfx plugin control over HLE/LLE would that be possible? First the HLE display list is sent to the gfx plugin. Then the plugin can decide if it wants to HLE the dlist or send it back to the RSP. If it is sent back to the RSP then the plugin receives low level dlists.

This would make N64 emulation much easier. If a microcode is not implemented then LLE could be used automatically.

**Comment by LegendOfDragoon** [comment]
>Plus, to play devils advocate, does it matter if the soft RDP is LLE only if the RSP recompiler covers both?

LLE is too slow.

>I think it would be better to give the gfx plugin control over HLE/LLE would that be possible? First the HLE display list is sent to the gfx plugin. Then the plugin can decide if it wants to HLE the dlist or send it back to the RSP. If it is sent back to the RSP then the plugin receives low level dlists.
This would make N64 emulation much easier. If a microcode is not implemented then LLE could be used automatically.

I'm not sure if that's necessary, given the fact that Project64 already has per game settings for selecting HLE or LLE.

**Comment by purplemarshmallow** [comment]
>I'm not sure if that's necessary, given the fact that Project64 already has per game settings for selecting HLE or LLE.

But the core does not know what a plugin supports. If a plugin does not support HLE it will not work without manual settings. Like angrylion's plugin. Also if someone wants to reverse an unknown ucode it would make things easier.

Automatic detection is usually better than manual per game settings. It should be more flexible. I think it would be useful. Should be possible to get rid of all unknown ucode errors.

**Comment by purplemarshmallow** [comment]
I made an experiment. I built angrylion's plugin into z64gl and used it for framebuffer emulation. It's a bit faster than angrylion's plugin but not so much. It can be more optimized because a lot of work is done twice. One problem is the plugins can't share the same TMEM because z64gl's TMEM emulation is severely incorrect.

Here's a build if someone is interested.
https://drive.google.com/file/d/0B7Y6r4SpC_QQdzVuWURyMmkxX2M/view?usp=sharing

The big disadvantage is how do you know if you should render a buffer in software or not? I think it's impossible. With this build everything is rendered in software. This makes things slow.

**Comment by LegendOfDragoon** [comment]
Weird. For me, it crashes on pj64. On 1964 it works, but the graphics are completely messed up (but works if i use mesa).

>But the core does not know what a plugin supports.
Good point.

**Comment by fallaha56** [comment]
@purplemarshmallow cool effort :) sadly completely crashing for me too in PJ64/Win10/nVidia :(

@gonetz shame the core and plugins aren't a little more integrated to make all this easier, at least the HLE/LLE settings

@LegendOfDragoon i know what you're saying about LLE and speed but it just seems to be necessary for accurate emulation and the existing software RSP and RDP seem to have a lot of scope optimisation

**Comment by LegendOfDragoon** [comment]
>i know what you're saying about LLE and speed but it just seems to be necessary for accurate emulation...plus the existing software RSP and RDP seem to have a lot of scope for optimisation

It's easier said than done, to actually speed it up to satisfactory levels. Especially since performance isn't the only problem.

**Comment by purplemarshmallow** [comment]
z64gl hardly causes any slowdown it's very well optimized. I think I found why it is so fast. Looks like it has some sort of delayed rendering. It collects data and then renders everything at once which is faster but I'm not 100% sure. Almost all slowdown comes from the RSP.

I'm not sure why it crashes. It works for me. PJ64/Win10/nVidia  here as well. Mario Tennis looks good with software fb. Body Harvest collisions also work. And OOT subscreen delay is gone. But it's slow.
![1](https://cloud.githubusercontent.com/assets/7283660/12007733/50d3d5c4-ac12-11e5-9b68-2ad0918bfd32.jpg)

**Comment by purplemarshmallow** [comment]
I managed to modify angrylion's plugin to render only depth values. This is a lot faster z64gl still runs fullspeed in some games on my system. And I get pixel accurate depth! Color rendering is much slower. It's possibly faster to copy data from the GPU.

GLideN64 can use angrylion's plugin as well for depth rendering but it will only work in LLE.

**Comment by fallaha56** [comment]
hi @LegendOfDragoon as you know, i'm no great programmer so please take anything i say with a slight pinch of salt!

nonetheless, MAME literally speeds up 400-500% when multi-threading the 3Dfx; Gauntlet Legends goes from an unplayable 20fps to 100% playable

no need to re-invent the wheel, can/can't this approach be duplicated? imitation=sincerest form flattery ;)

same for the software VI filter that costs 10+fps yet is totally do-able (if not already done?) in OpenGL?

**Comment by fallaha56** [comment]
@purplemarshmallow can you remind me what config files etc i need for z64gl? not sure i installed it last time i re-did my rig

**Comment by purplemarshmallow** [comment]
you put glew32.dll and SDL.dll in the emulator's directory. It's important that you have the right version I included them. and you put z64gl.dll and z64gl.conf in the plugin directory. Here's my config file:
https://drive.google.com/file/d/0B7Y6r4SpC_QQV3FlZDdONDYxa2s/view?usp=sharing

**Comment by LegendOfDragoon** [comment]
I have my doubts about z64gl being well optimized, unless the performancr issues I encountered are caused by bugs. Some games are fast, like sm64, but some games run too slow, even when compared to angrylions, such as vigilante 8

@fallaha56 3Dfx is different than N64. Last I checked, MESS for N64 was not full speed. I doubt it would be feasible to split the rdp into multiple threads. As for using opengl for vi filters, I'm not so sure it will be that much faster. Besides, I dont think you should worry about optimizing filters until the RDP runs beyond satisfactory speed.

**Comment by purplemarshmallow** [comment]
>I have my doubts about z64gl being well optimized, unless the performancr issues I encountered are caused by bugs. 

I think it's caused by bugs. It's based on a very old version of the MESS renderer. Many things work incorrect. If a game works correctly then it's very fast. Like OOT or World Driver Championship.

**Comment by LegendOfDragoon** [comment]
>I think it's caused by bugs. It's based on a very old version of the MESS renderer. Many things work incorrect. If a game works correctly then it's very fast. Like OOT or World Driver Championship.

Odd, I get full speed in OOT, but not WDC with z64gl. Anyway, does your experimental build implement coverage?

**Comment by purplemarshmallow** [comment]
>Odd, I get full speed in OOT, but not WDC with z64gl

Can you run GLideN64 LLE? How fast is it compared to z64gl?

>Anyway, does your experimental build implement coverage?

Framebuffers rendered by angrylion's plugin are rendered with correct pixel coverage emulation. Framebuffers rendered by z64gl not.

@gonetz should work on a PR and replace the (not properly working) LLE depth buffer to Rdram code with software depth buffer rendering. That's easily possible with angrylion's plugin.

**Comment by LegendOfDragoon** [comment]
>Can you run GLideN64 LLE? How fast is it compared to z64gl?

Unfortunately, I cannot run GlideN64 properly. I only have OpenGL 2.1. What I usually do is have other people to test for me, when I'm curious about GLideN64 :smile: . For instance, if I remember correctly, Mortal Kombat 4 ran really slow with GLideN64 LLE for multiple people, but z64gl should be full speed.

>@gonetz should work on a PR and replace the (not properly working) LLE depth buffer to Rdram code with software depth buffer rendering. That's easily possible with angrylion's plugin.

I think that's a great idea. Angrylion's code is really nice.

**Comment by mudlord** [comment]
So hybrid depth emulation? Nice solution.

Although, isn't there the pressing problem of the LLE triangle function being incorrect?

**Comment by purplemarshmallow** [comment]
The LLE triangle function is not perfect. But I don't think that's the biggest problem. The biggest problem is speed. z64gl mostly runs fullspeed on my system in Zelda OOT with software depth buffer. In some places it slows down. GLideN64 LLE is slower.

**Comment by mudlord** [comment]
Optimize RSP with proper dynarec?

**Comment by purplemarshmallow** [comment]
That would be good. PJ64's RSP recompiler is about as fast as cxd4's interpreter. Also it's not compatible with some games. cxd4 did a good job with optimizing his interpreter. But a recompiler can be much faster

**Comment by LegendOfDragoon** [comment]
>PJ64's RSP recompiler is about as fast as cxd4's interpreter.

When you're using a fast LLE graphics plugin, you can usually see a clear difference in speed between PJ64's RSP Recompiler and cxd4's RSP.

>Also it's not compatible with some games.

Aside from 3 factor 5 games, what others are not compatible?

**Comment by purplemarshmallow** [comment]
>When you're using a fast LLE graphics plugin, you can usually see a clear difference in speed between PJ64's RSP Recompiler and cxd4's RSP.

Yes the RSP recompiler is faster

>Aside from 3 factor 5 games, what others are not compatible?

I don't know any other games

**Comment by purplemarshmallow** [comment]
I implemented software depth buffer rendering in GLideN64. My implementation works like this: First dlists are sent to angrylion's plugin. Then they are executed but many commands are ignored like RDP_TriTxtr, RDP_TriFill, ... Then they are sent to GLideN64 and executed again. Possibly this implementation can be improved.

The problem is that angrylion's plugin still draws something in the color buffer. It's mostly white with some gray areas. This conflicts with GLideN64's validity checking implementation and buffers are removed from the GPU.

Maybe angrylion's plugin can be optimized for rendering depth values only.

**Comment by purplemarshmallow** [comment]
I experimented with copy from RDRAM and software depth buffer rendering

This is the depth image. It's used as background in the subscreen. Usually the game renders the subscreen background in the depth buffer area but color rendering is disabled.
![3](https://cloud.githubusercontent.com/assets/7283660/12012357/aafe1b5c-acf0-11e5-8235-ed7a56e606cf.jpg)

angrylion's plugin draws this in the color buffer area if commands for color rendering are not executed
![1](https://cloud.githubusercontent.com/assets/7283660/12012364/fca47b9a-acf0-11e5-8015-8306599215c5.jpg)

And this is the pixel coverage? Why does this show up if I open the subscreen? Looks like the line mode in Goldeneye #226 
![2](https://cloud.githubusercontent.com/assets/7283660/12012371/2ca7c112-acf1-11e5-998c-045299165c20.jpg)

**Comment by AmbientMalice** [comment]
Pixel coverage emulation will hopefully fix "soft" shadows in Turok 3/Jet Force Gemini in conjunction with the alpha fix Gonetz reverted due to it creating fuzzy black crawling shadows. Plus of course pixel coverage seems to be used for a lot of advanced effects.

This is all very promising. If this actually works, it will be a dramatic accuracy and compatibility improvement. But I am concerned about the performance impact of software depth buffering.

**Comment by purplemarshmallow** [comment]
>Pixel coverage emulation will hopefully fix "soft" shadows in Turok 3/Jet Force Gemini in conjunction with the alpha fix Gonetz reverted due to it creating fuzzy black crawling shadows.

There has been progress. The fb info branch supports 8 bit buffer copy. If you enable copy auxiliary in Jet Force Gemini the shadows will look correctly with soft edges.

>But I am concerned about the performance impact of software depth buffering.

Me too. In LLE performance is most important. I think my implementation can be improved to speed things up. Depth buffer copies from the GPU with emulator notification will likely always be faster. Also angrylion's plugin does not work in HLE.

I think there is still this problem with the incorrect depth format. GLideN64 uses float instead if int. Precision is lost.

**Comment by LegendOfDragoon** [comment]
>GLideN64 uses float instead if int. Precision is lost.

Why does it use floating point? That's also one of the reasons I haven't bothered with working on z64gl.

**Comment by purplemarshmallow** [comment]
Int depth fromat was removed from GLideN64 because it was not working properly. Float worked better for unknown reasons. Not sure what it has to do with z64gl

**Comment by LegendOfDragoon** [comment]
>Int depth fromat was removed from GLideN64 because it was not working properly. Float worked better for unknown reasons. Not sure what it has to do with z64gl

I see. I brought up z64gl because it uses floating point too. Idk of any hardware plugin that doesn't do float. For that reason, I'm sticking with software rendering for now.

**Comment by mudlord** [comment]
well too bad, modern hw now supports int now.

**Comment by fallaha56** [comment]
in case it helps i think this integer problem was solved in Dolphin for DX/OpenGL:
https://dolphin-emu.org/blog/2014/03/15/pixel-processing-problems/

@LegendOfDragoon also if it helps i may have a OpenGL4 GFX card to donate to the cause at some point this year...

**Comment by LegendOfDragoon** [comment]
@fallaha56 Thanks for the offer, but I use a laptop, so I cannot just plug in a new gfx card. Besides, I don't think I will be interested in OpenGL anytime soon. I have yet to really learn any graphics api. I mostly focus on learning cpu rather than gpu.

**Comment by purplemarshmallow** [comment]
>Optimize RSP with proper dynarec?

@mudlord I'm not sure anymore if that's the best thing to do. Maybe it's possible to reproduce the RSP's work on the GPU and process dlists on the GPU. Should be possible to make a fast LLE hardware renderer this way.

**Comment by LegendOfDragoon** [comment]
>@mudlord I'm not sure anymore if that's the best thing to do. Maybe it's possible to reproduce the RSP's work on the GPU and process dlists on the GPU. Should be possible to make a fast LLE hardware renderer this way.

I could be wrong, but I have a feeling that RSP instructions do not map too well to the GPU. I think that's better suited for HLE because it's more direct. In most games I've played / tested (a few dozen games),  the RSP was not the main bottleneck, so the primary focus should be optimizing the graphics plugin. A decent RSP dynarec will get good enough speeds for most games.

But yes, I don't think dynarec is the "best" option. Just good enough to get the job done in a reasonable amount of time.

**Comment by purplemarshmallow** [comment]
The RSP and the RDP do the rendering. The RDP is already emulated in hardware. Maybe it's possible to move rendering tasks from the RSP to the GPU to speed things up. But if it works too different from PC hardware it can be difficult.

**Comment by fallaha56** [comment]
Is all this where OpenCL/heterogeneous compute could come in? 2.1 is basically C++ i thought:
http://www.anandtech.com/show/9039/khronos-announces-opencl-21-c-comes-to-opencl

and the Intel IRIS GPUs can do some cool shared memory tricks with their cache N64-style:
https://software.intel.com/sites/default/files/managed/9a/77/instantaccess-tiling2.pdf

-we certainly aren't going to get much in the way of simple clock speed increases from the next generation of CPUs!

**Comment by LegendOfDragoon** [comment]
>The RDP is already emulated in hardware.

Yes, but it still appears to be inefficient. When I look at the profiler while using a HW rendering plugin, the vast majority of the bottleneck is D3D or OpenGL, depending on the plugin I'm profiling. I wish I knew a good profiler to use for OpenGL and D3D, so that I could figure out how to speed it up.

**Comment by fallaha56** [comment]
is this any use?
https://software.intel.com/en-us/gpa

surely if the bottleneck is OpenGL/D3D then either Vulcan or DX12 has to be part of the answer though?

at the risk of sounding like a broken record, there's a DX12 build of Dolphin that shows 150-160% speedups in GPU-heavy games:
https://forums.dolphin-emu.org/Thread-unofficial-dolphin-dx12-backend

**Comment by purplemarshmallow** [comment]
>Yes, but it still appears to be inefficient. When I look at the profiler while using a HW rendering plugin, the vast majority of the bottleneck is D3D or OpenGL, depending on the plugin I'm profiling

You should not forget that HLE skips RSP emulation. That's the reason why it's so fast. But if you add RSP emulation it slows things down. So the key is to make the RSP and RSP - RDP communication more efficient

**Comment by LegendOfDragoon** [comment]
>is this any use?
https://software.intel.com/en-us/gpa

Dunno, but I may look into it one day :smile: .

>surely if the bottleneck is OpenGL/D3D then either Vulcan or DX12 has to be part of the answer though?
at the risk of sounding like a broken record, there's a DX12 build of Dolphin that shows 150-160% speedups in GPU-heavy games:

I'm sure the performance would likely be significantly better with either of those 2 API's but it kind of defeats the purpose (for me anyway). If I were to release software, I would want it to be usable for most users, so it has to be compatible with a decent amount of hardware. If I wanted to just take the easy route, I could just buy a skylake (once I finally save up enough money) and call it a day, because at that point, I'd be able to run practically any game i care about, with full speed on a software renderer.

>You should not forget that HLE skips RSP emulation. That's the reason why it's so fast. But if you add RSP emulation it slows things down. So the key is to make the RSP and RSP - RDP communication more efficient

I'm aware of the key difference between HLE and LLE. I should have mentioned that I also profiled Jabo 1.7 LLE (just to see what dll's were the bottleneck) and z64gl as well. Even HLE is slow in some cases. For instance, on Glide64Final, I don't get steady 60 fps in Dark Rift. SD Hiryuu is another great example of inadequate speed in HLE. I do believe some games may need HLE to run at good speeds, such as Vigilante 8, but most games should run fine in LLE.

**Comment by purplemarshmallow** [comment]
I believe triangle commands can be more optimized. If there's too much geometry it slows things down. I tested Zelda OOT and z64gl. If I stand in one corner of the map and press c up and look against a wall I can easily run it at more than 200% speed. But if I turn around and look into the scene it cuts speed in half. I think there is rendering work that must be offloaded to the GPU

I'm not sure how much DX12/Vulcan can do. In dolphin it speeds up GPU heavy games. I think in N64 emulation the bottleneck is mostly the CPU

**Comment by FishamanP** [comment]
Just my two cents, DX12/Vulcan improve GPU performance because they make the CPU more efficient at feeding the GPU. They give the most gains on slow CPUs paired with strong GPUs.

I think Vulcan is a better route than DX12 because of its portability, but I don't know if it will even help. The biggest single performance boost is from multithreading the command lists.

**Comment by fallaha56** [comment]
It's the setup aspect/triangle thruput that Vulvan/DX12 accelerates.

Guess like you say trick will be coding it for GPU, that said again those APIs are very programmable/multithread-able ;)

**Comment by purplemarshmallow** [comment]
This sounds good would be nice to have a Vulvan/DX12 renderer optimized for LLE triangles. My GPU does not support it unfortunately I can't test DX12 dolphin. But they are planning to release more drivers for older hardware so system requirements won't be that high.

**Comment by fallaha56** [comment]
@purplemarshmallow this is where it gets interesting i think -any recent GPU will support the polygon speedups, not just Maxwell or Fury (I checked and in fact this is what the DX12 Dolphin does, the speedup is entirely polygon throughput related, no special DX12 features are used)

**Comment by LegendOfDragoon** [comment]
>But they are planning to release more drivers for older hardware so system requirements won't be that high.

That's my main issue. Seems almost pointless if only new hardware can utilize it. New hardware should already run these games at full speed. If I had the hardware capable of using it, I'd still be interested in the API, but it shouldn't be the main focus if the goal is to improve emulation for everyone.

**Comment by mudlord** [comment]
@fallaha56 Vulkan is not a silver bullet. It will involve a complete rewrite to what is there currently. 
You cannot just drop it in with OGL components.

**Comment by purplemarshmallow** [comment]
>That's my main issue. Seems almost pointless if only new hardware can utilize it. New hardware should already run these games at full speed.

To run most games at fullspeed with angrylion's plugin you will need a 4Ghz skylake I guess. DX12 compatible hardware is not fast enough in most cases.

**Comment by fallaha56** [comment]
@mudlord i know enough of your expertise to know not to question it!

i'm merely suggesting that it feels like we've reached the limit of what's achievable on older hardware -admittedly i do this from the perspective of an experienced system builder rather than that of a programmer...

without a multi-threaded, SSE4/AVX +/- OpenCL approach how else can N64 emulation be truly achieved? -i mean that as a genuine question from a non-N64-expert (me) to the experts (you guys)

otherwise isn't there a risk of wasting endless effort? processors are not getting faster so much as smarter...

shouldn't folk be pooling their efforts? -i don't know a single emu with as many plugins as Project64(!)

(also, after sentimentally struggling on for years with a Core2Duo HP touchsmart, i finally caved and bought a Surface4Pro -i'm now running F-Zero GX in Dolphin@60fps...on a 15W Ultrabook...i couldn't do that in DX11, just wish i'd done it sooner...)

**Comment by LegendOfDragoon** [comment]
>To run most games at fullspeed with angrylion's plugin you will need a 4Ghz skylake I guess. DX12 compatible hardware is not fast enough in most cases.

I was exaggerating a bit, but in this case was more thinking about hardware rendering plugins. Aside from the awful LLE performance in some games for GLideN64, what games are not full speed currently in GLideN64?


>shouldn't folk be pooling their efforts? -i don't know a single emu with as many plugins as Project64(!)

Idealy, perhaps. But it doesn't work that way. Who wants to use my assembly code? Nobody I know, besides like 1 person :smile: . Diversity is good anyway, because it leads to more bug fixes. There are bugs that can be revealed after changing other parts of code. Even different compilers can expose bugs.

Then, there's the fact that different people have different opportunities and goals. Would I be more inclined to learn and use Vulkan over D3D11, if I had the hardware support? Absolutely, but I do not, so I have to settle for what I got.

>i'm merely suggesting that it feels like we've reached the limit of what's achievable on older hardware -admittedly i do this from the perspective of an experienced system builder rather than that of a programmer...

Absolutely not! A big problem is the software, not the hardware. I honestly got sick of people telling me the reason I'm not getting good performance is because "you're running on a toaster", yet other software (sometimes even more accurate), run better.

>without a multi-threaded, SSE4/AVX +/- OpenCL approach how else can N64 emulation be truly achieved? -i mean that as a genuine question from a non-N64-expert (me) to the experts (you guys)
otherwise isn't there a risk of wasting endless effort? processors are not getting faster so much as smarter...

By writing more efficient code, rather than relying on newer API's and hardware. I think it's best to set reasonable limits, rather than extreme. I don't think you should raise the requirements, until you've pushed it to the limit and realize that it just won't cut it.

**Comment by purplemarshmallow** [comment]
>Aside from the awful LLE performance in some games for GLideN64, what games are not full speed currently in GLideN64?

In HLE it depends on your settings. If you enable per-frame copy color, copy depth, noise, N64 depth compare it can slow things down below fullspeed even with HLE.

Vigilante 8 needs both copy color and copy depth and it's very slow in this game because there are multiple dlists per frame. Emulator notifications can help. Star Wars EP1 Racer is slow because of copy depth and noise. N64 depth compare can severely slow down any game.

>By writing more efficient code, rather than relying on newer API's and hardware. I think it's best to set reasonable limits, rather than extreme. I don't think you should raise the requirements, until you've pushed it to the limit and realize that it just won't cut it.

Glide64 was pushed to the limits. To emulate the combiner properly you need modern GLSL. For N64 depth compare you need image textures. For efficient copy color you need glBlitFramebuffer. There is a reason why GLideN64 raised the hardware requirements. If there is a new technology that can be useful why not use it?

Also supporting very old hardware can be a massive overhead and you have to question is it worth it? Rice Video has 6 different combiners and what does it bring today?

**Comment by LegendOfDragoon** [comment]
>In HLE it depends on your settings. If you enable per-frame copy color, copy depth, noise, N64 depth compare it can slow things down below fullspeed even with HLE.

Oh wow, I figured high end machines would at least run the games full speed. Guess that's not always the case.

>Glide64 was pushed to the limits.

I beg to differ. Certainly not in performance or accuracy. Sure there are API limitations, but the rest of the code was not the best it could be. Not only could it be more optimized, the HLE code could also improve in accuracy.

>If there is a new technology that can be useful why not use it?

There's nothing wrong with using newer technology, as long as you don't rely on it when it's not necessary.

>Also supporting very old hardware can be a massive overhead and you have to question is it worth it? Rice Video has 6 different combiners and what does it bring today?

It's still useful for people with older hardware. Even I still use it for some games, due to its amazing performance. I appreciate having plugins that I can actually use and play with.

**Comment by purplemarshmallow** [comment]
If you support older hardware you have to question how much effort does it create? How many people still use it? Does it harm me elsewhere (porting to OpenGL ES)? How much extra code is required? Does it prevent progress (relying on deprecated functionality)? Then you have to draw a line somewhere.

Rice's OpenGL 1.1 mode is not useful. OpenGL 1.1 just can't do the job.

**Comment by LegendOfDragoon** [comment]
>If you support older hardware you have to question how much effort does it create? How many people still use it?

These are good questions. It is important to think about this.

>Then you have to draw a line somewhere.

Absolutely. I just don't like the idea of upgrading requirements just to take the easy way out. For example, SSE2 is pretty much good enough for the RSP, so I think there is no good reason to make a RSP that requires SSE4 as the minimum. For the RDP, I don't think AVX would be that beneficial so I'd say the same thing for that. There's no good reason to have AVX as the minimum requirement for the RDP. People think multi-threading will make Angrylion's run full speed, but Shunyuan already tried that and failed miserably.

>Rice's OpenGL 1.1 mode is not useful. OpenGL 1.1 just can't do the job.

I agree that OGL 1.1 cannot do the job. However, OGL 2.1 can do a pretty good job for many games. z64gl looks amazing. My issues with it are bugs and performance, which could both be improved without upgrading API's. I'd like to at least support hardware that is 8 or 9 years old, even if it doesn't fully support every feature. Considering that other code bases already use older versions of OpenGL, I do not think it would be too much work. My comment was mainly towards the idea of upgrading API's to just magically improve the results, without putting much work into actually improving the code.

**Comment by fallaha56** [comment]
@LegendOfDragoon 'pretty good job for many games' is pointless IMHO...you're doomed to failure in this scenario

plus SSE<b>4</b> is now over 8 years old BTW...and the next Intel HEDT will have 10 native cores

IMHO Dolphin utterly and completely did the right thing ditching Win32 and DX9, so did @gonetz by aiming for an OpenGL level that would offer 100% functionality

after all, no one uses UltraHLE any more!

**Comment by objectchris** [comment]
I think the emulation hardware scene has changed quite a bit over the last few years. It seems like an increasing amount (majority?) of users use low spec embedded systems (eg raspberry pi) or mobile/tablet platforms. SSE or DX are rarely available in such platforms (although they are increasingly multi-core).

**Comment by purplemarshmallow** [comment]
z64gl can be improved without updating APIs. Why is this doomed to failure?
TMEM load commands in z64gl are 100% copy paste from Ville Linde's MESS renderer. This code is incorrect and causes bugs slowdowns and crashes in many games. It can be replaced with current MAME code or angrylion's code I tried both. The plugins have the same structure. The problem is that there are wrong assumptions in the texture code and while some slowdowns and crashes are gone many textures will turn into garbage. The texture code needs to be corrected as well but I doubt this can be done with a simple copy/paste.

**Comment by AmbientMalice** [comment]
The irony here is that I've been asking for some time why nobody bothered cleaning up z64gl instead of making yet another HLE plugin. GLideN64 has made a lot of improvements when it comes to mipmapping, framebuffer, etc, but basing it on GLN64 felt quite shortsighted, especially since N64 HLE video is problematic due to missing and incomplete ucodes that will likely never be fixed.

It's really not ideal that GLideN64's depth buffer has turned out to be such a mess. Scrapping it and going for a software approach will help, that's a lot of work down the drain.

**Comment by FishamanP** [comment]
First, it's important to define what the main goals are here. 100% accuracy, or the "good enough" 97%? Portability through the future, or speed now? These forward-thinking choices seem like the goals you have in mind, and I think that merits forgetting hacks to speed up older hardware. Even a pure C89 implementation will eventually get full speed on embedded hardware, as time marches on and tech improves. That's not to say you can't write optimized code, but you have to optimize from a strong foundation.

**Comment by purplemarshmallow** [comment]
>It's really not ideal that GLideN64's depth buffer has turned out to be such a mess. Scrapping it and going for a software approach will help, that's a lot of work down the drain.

Software depth buffer rendering can't help with N64 depth compare emulation on the GPU. It can't help with "game copies color buffer into depth buffer area" scenarios.

Software depth buffer rendering won't magically solve all problems. It's one way to provide the CPU with data but it can't help with anything else. Another way to provide the CPU with data is copy depth with emulator notification this can be very fast. Depth buffer emulation on the GPU can be improved. I think using integer math for depth calculations on the GPU should also give accurate results.

**Comment by LegendOfDragoon** [comment]
>@LegendOfDragoon 'pretty good job for many games' is pointless IMHO...you're doomed to failure in this scenario

It's better than not being able to play games at all.... If I'm able to play most of the games I want at a satisfactory level (which is already almost the case thanks to the hard work of other developers as well as myself), how is that doomed to failure? I don't think you fully understand the situation. If you want perfect accuracy, then software rendering is the only realistic option.

@fallaha56 I'd really like to know why it's pointless to you.

>plus SSE4 is now over 8 years old BTW...

That's fine and I have no problem using SSE4 to optimize the RDP. However, I think it would be sad to not to spend an extra hour or two, to ensure SSE2 works...

>TMEM load commands in z64gl are 100% copy paste from Ville Linde's MESS renderer. This code is incorrect and causes bugs slowdowns and crashes in many games. It can be replaced with current MAME code or angrylion's code I tried both. The plugins have the same structure. The problem is that there are wrong assumptions in the texture code and while some slowdowns and crashes are gone many textures will turn into garbage. The texture code needs to be corrected as well but I doubt this can be done with a simple copy/paste.

That is very interesting. Maybe I should look into this. Can you provide a link to Ville Linde's code? I want to look at the similarity so I can see if I can fix it :smile: .

>The irony here is that I've been asking for some time why nobody bothered cleaning up z64gl instead of making yet another HLE plugin. GLideN64 has made a lot of improvements when it comes to mipmapping, framebuffer, etc, but basing it on GLN64 felt quite shortsighted, especially since N64 HLE video is problematic due to missing and incomplete ucodes that will likely never be fixed.

At one point, I did want to make z64gl better, but it was difficult for me to get started with it. It was much easier for me to just learn from Angrylion's code. I do feel that z64gl has a lot of potential.

**Comment by purplemarshmallow** [comment]
Here is the commit where I replaced the rdp_load_block implementation with current MAME code https://github.com/purplemarshmallow/z64/commit/eeab785149c14c6e9c3564f7cfb0659825827554
It fixes menus and backgrounds in F-Zero X but breaks textures ingame. I also tried angrylion's code it gives the same results. Then I replaced the code in the old MESS renderer and also got the same texture corruption. I think this proves that there are wrong assumptions elsewhere

The old MESS renderer is included with z64gl. It's named rdp-mess.cpp and you can even compile a zilmar spec plugin from it. z64gl is based on that.

**Comment by LuigiBlood** [comment]
So to sum it up, RDP LLE is slow because of RSP being too slow?
Also I tried z64gl with angrylion code build with Mario Artist Talent Studio and Polygon Studio: both games are totally functional. (Paint Studio = black screen on menu, for some reason though)

**Comment by LegendOfDragoon** [comment]
>The old MESS renderer is included with z64gl. It's named rdp-mess.cpp and you can even compile a zilmar spec plugin from it. z64gl is based on that.

Sweet! I'll take a look at the code today and see if I can fix anything. Any idea why z64 crashes though? And for me, it creates a new window, which is kinda weird xD. Maybe I'll just backport Angrylion's DDRAW code xD.

**Comment by theboy181** [comment]
AIO "legendofdragoon" and I spent a few hours testing rsp's with the no video plugin while watching the % usage on PJ and the rsp actually was not as slow as the rdp in all but a few rare cases. 

When using optimized rsp that gap is even smaller. We need to stop asuming that the issue with the speed of the rcp is the rsp's fault. 

I recommend we continue to test with no video plug and document results. 

I like the sound of z64gl with updated angrylion code. What can I do to help?  Will it compule in MSVS?

**Comment by LegendOfDragoon** [comment]
Ok, after porting some basic VI code, I got z64 to work. I'm wondering if I did something wrong or if it's really this buggy!

What confuses me is that in Super Smash Bros, shields and items are messed up in z64gl, but if I use mesa3d they are fine. Yet in z64 I also see the same messed up shields and items. This could be fun :smile: .

**Comment by theboy181** [comment]
Have you tried to install the manufactures video drivers for your laptop to see if your gl issues go away?

**Comment by LegendOfDragoon** [comment]
>Have you tried to install the manufactures video drivers for your laptop to see if your gl issues go away?

Last time I checked, I already had latest drivers. At this point, I'm convinced that the bigger problem is the code itself, rather than hardware / drivers. Same goes for the other OpenGL plugins.

**Comment by fallaha56** [comment]
@LegendOfDragoon before i say anything else, i want to say i think it's cool to see the passion here

to answer your question, why 'pointless': i'm a hardware guy, IMHO it's pointless to fight the hardware cycle

also IMHO certain hardware jumps are key and must be embraced: i have a 5 y/o CPU but a 1 y/o GFX card (and no, it's not a Titan X)

if someone could write a single-threaded, fast'n'accurate, SSE2/OpenGL2 N64 emu they'd have done it by now

what makes me feel sure about this? i can barely program so i just look at which emus are the most successful/popular just now, Dolphin and PCSX2 -they drop legacy rapidly, have 1-2 plugins tops and (most importantly) collaborate

**Comment by fallaha56** [comment]
as always, wishing to be constructive in the long run, happy to help/contribute any way i can...just not sure what that is!

**Comment by LegendOfDragoon** [comment]
>to answer your question, why 'pointless': i'm a hardware guy, IMHO it's pointless to fight the hardware cycle

That sounds overly pessimistic. Not sure why you would think it's pointless to fight the hardware cycle.

>also IMHO certain hardware jumps are key and must be embraced

Sounds reasonable, but it is debatable where to draw the line.

>if someone could write a single-threaded, fast'n'accurate, SSE2/OpenGL2 N64 emu they'd have done it by now

One, you're assuming enough people even tried. Two, you're assuming we've reached the limit in accuracy and performance. Both of which are untrue. You didn't even mention 32bit which is a significant obstacle for the RDP's performance. If I had that mentality, I would have never bothered working on N64 emulation. If i didn't think I could do better, why even try lol?

>what makes me feel sure about this? i can barely program so i just look at which emus are the most successful/popular just now, Dolphin and PCSX2 -they drop legacy rapidly, have 1-2 plugins tops and (most importantly) collaborate

What you have to understand is that every situation is unique. What people do for other emulation systems is not 100% relevant for N64 emulation. Tbh, popularity does not matter much. What's popular isn't always the best. Besides, it would make more sense in general for newer consoles to require newer hardware to emulate properly.

The real question is, what standards do you go by for "accurate"> If your standards are very high, then I'm not sure hardware rendering is ideal for you.

**Comment by purplemarshmallow** [comment]
>AIO "legendofdragoon" and I spent a few hours testing rsp's with the no video plugin while watching the % usage on PJ and the rsp actually was not as slow as the rdp in all but a few rare cases.

>When using optimized rsp that gap is even smaller. We need to stop asuming that the issue with the speed of the rcp is the rsp's fault. 

Sure it's the RSP's fault. I tested with no video plugin. In Rogue Squadron (ingame) I get:
13%R4300i 70%GFX 17%Alist ~70VI/s

So the RSP eats 70% performance for GFX emulation
Now I tested with z64gl, same point in game
10%R4300i 77%GFX 13%Alist ~55VI/s

To sum it up the RSP eats 70% z64gl eats 7%. All slowdown comes from the RSP. I tested with cxd4's interpreter it's the fastest RSP plugin for this game. PJ64's interpreter is slower.

>Ok, after porting some basic VI code, I got z64 to work. I'm wondering if I did something wrong or if it's really this buggy!

I guess you did right and it's really this buggy. This code was the first LLE GFX emulation attempt. Impressive for its time but now it's very bad that z64gl is based on something that incorrect. Now the RDP is fully reversed and there is the chance to make it better.

**Comment by theboy181** [comment]
You used a game that is heavy with audio rsp instructions. I think that can be optimized rather quickly. 

Test with games that are not heavy audio. So all boss and musyx games off the table.

**Comment by purplemarshmallow** [comment]
Super Mario 64
7.9%R4300i 83.8%GFX 8.3%Alist ~115VI/s

no video plugin, cxd4's RSP. 83.8% is even more extreme

also audio RSP instructions should be under Alist not GFX

**Comment by theboy181** [comment]
Maybe AIO can chime in on this. I think you may need a specific no video plugin that supports lle. Cause your results are different than mine.

**Comment by fallaha56** [comment]
@LegendOfDragoon 

re:accurate:
a 64bit multi-threaded N64 emu running OpenGL4+ is going to be accurate'n'fast enough for pretty much anybody so why not make that the baseline? a good software renderer is always needed for reference anyway, so keep one around too

re: pointless:
i'm not pessimistic but realistic ;) -look at the once mighty Kodak, where are they today for fighting digital photography/progress? bankrupt lol!

re: 'the line', it's drawn by performance jumps:
my CPU's an old i7 980x, off eBay for <$250, OC to 4.2Ghz; this oft matches any Skylake, GPUs are a different story, performance firmly with Moore's law

re: 64bit:
i'm not assuming people haven't tried! i'm saying since it hasn't happened that a different approach is required; i've had several tiring conversations esp with HatCat who thinks he'll make some mythical 32bit SSE2 single-threaded 100% accurate RSP sometime before the year 2025, if that happens i'll buy him a new hat and eat mine!

re: unique
i understand each is different but again if it was possible/practical without 64bit, multicore etc it would have happened by now; popularity is in part definitely one measure of how accurate and fast an emu is, we all love Mario but there's no fun playing him if he's going to crash and glitch every other frame...

just my 2 1/2 cents...

**Comment by LegendOfDragoon** [comment]
>Sure it's the RSP's fault. I tested with no video plugin. In Rogue Squadron (ingame) I get:
13%R4300i 70%GFX 17%Alist ~70VI/s

Rogue Squadron uses the RSP more than most games. Also, cxd4's RSP isn't particularly optimized for the game either.

Part of the problem is that on different hardware, people will get different performance for hardware rendering plugins. However, when comparing Angrylion's RDP performance to RSP, I can definitely say RDP is more of a bottleneck for most games. For me, some games are so bad on z64gl and Jabo's that I'm better off using a software plugin. It makes me question what is going on. I really should look into profiling gfx API code.

>I tested with cxd4's interpreter it's the fastest RSP plugin for this game. PJ64's interpreter is slower.

I should try to fix the factor 5 issues in PJ64's RSP recompiler soon. Maybe this weekend, if I get around to it.

>I guess you did right and it's really this buggy. This code was the first LLE GFX emulation attempt. Impressive for its time but now it's very bad that z64gl is based on something that incorrect. Now the RDP is fully reversed and there is the chance to make it better.

I see. I hope others will be interested in fixing z64gl. I'm a little more motivated, but it's still a lot of work. I read through it and saw that it's quite different than Angrylion's. My issue is that I'm not good with gfx API's ;/ . So I won't know how to translate the code to OpenGL.


>re:accurate:
a 64bit multi-threaded N64 emu running OpenGL4+ is going to be accurate'n'fast enough for pretty much anybody so why not make that the baseline? a good software renderer is always needed for reference anyway, so keep one around too

Again, it depends on standards. For cycle and pixel accurate, I do not believe that is enough. I'm not even sure if possible to get full speed and even pixel accurate on hardware rendering. I'd have to learn an API before figuring that out.


>re: pointless:
i'm not pessimistic but realistic ;) -look at the once mighty Kodak, where are they today for fighting digital photography/progress? bankrupt lol!

Your assumptions are not realistic. You really have to be a programmer who is involved with N64 emulation to get a more realistic idea of the situation. If I did not see how much potential there were for improvements, I would have never started working on N64 emulation.

>re: 'the line', it's drawn by performance jumps:
my CPU's an old i7 980x, off eBay for <$250, OC to 4.2Ghz; this oft matches any Skylake, GPUs are a different story, performance firmly with Moore's law

Even that isn't a clear line. For instance, the RSP is practically good enough with SSE2, but amazing with AVX2. The RDP is pretty much good enough with SSE4. The cpu emulation would be good enough with SSE2, etc. As for hardware rendering, I have no idea what a good limit would be.


>re: 64bit:
i'm not assuming people haven't tried! i'm saying since it hasn't happened that a different approach is required; i've had several tiring conversations esp with HatCat who thinks he'll make some mythical 32bit SSE2 single-threaded 100% accurate RSP sometime before the year 2025, if that happens i'll buy him a new hat and eat mine!

Actually, what I was saying is that there are not enough people trying. All the time, i hear certain people pinning their hopes on like 1 guy. That should tell you something. It shouldn't just be one or two people whom people are pinning their hopes on. This scene is not very active, contrary to other emulation scenes.


>re: unique
i understand each is different but again if it was possible/practical without 64bit, multicore etc it would have happened by now;

It can't happen if people never try in the first place... Less activity = less progress. That's the real problem here.

>popularity is in part definitely one measure of how accurate and fast an emu is, we all love Mario but there's no fun playing him if he's going to crash and glitch every other frame...

I beg to differ because the most popular != the best. There's a lot more to popularity than simply not crashing or glitching. Bias can play a huge role on what's popular. People often complain about N64 emulation, while using an improper setup. That should tell you something.

**Comment by gonetz** [comment]
>The RSP and the RDP do the rendering.

Depends on what you mean by "rendering". RSP prepares vertices for drawing: transformation, lighting, clipping. RDP does actual drawing. Texrect and Fillrect are RDP commands. Tri commands from HLE display list translated to calls of RDP Triangle command.

> should I work on a PR and replace the (not properly working) LLE depth buffer to Rdram code with software depth buffer rendering? That's easily possible with angrylion's plugin.

That depends on the goal you are trying to achieve. For example, you want to get pixel accurate depth buffer in RDRAM. If you use software render for depth buffer only, it will not be pixel accurate even with angrylion's plugin. Both color and depth buffer must be rendered in software to get pixel accurate result.

If your goal is to get depth buffer which is good enough for coronas emulation, software render is a good solution. However, angrylion's plugin is too complex and slow for that task, imo. Besides, it is other project code; you have to ask author’s permission for that. IMO, there is better alternative: Glide64 depth renderer:
* It is simple and fast. 
* It is compatible with HLE and LLE. 
* It is my code, so you don’t need to ask permission for using it in GLideN64.

**Comment by gonetz** [comment]
Regarding z64gl. I agree with purplemarshmallow: "z64gl hardly causes any slowdown it's very well optimized. … Almost all slowdown comes from the RSP." You hardly can make LLE emulation much faster by optimizing z64gl when maximal speed with no video plugin is ~70VI/s.

IMO, main problem of hardware LLE plugins is imperfect Triangle command. You may check how it is implemented in angrylion's plugin and compare with z64gl code. In fact, that is the work, which must be done first: understand line by line how Triangle command works and realize how to make correct implementation of it for hw plugin. This task is not easy.

**Comment by fallaha56** [comment]
@gonetz great to see you involved directly, perhaps you are the best person to answer my original question:

what do you think of the idea of plugging the entire Angrylion RDP code (but NOT the VI) into GLideN64 to create a highly-accurate semi-software plugin?

useful for reference purposes?

**Comment by gonetz** [comment]
What "semi-software" means? What should be rendered in software and what in hardware? How to mix the results?

IMO, hardware render can't be used for reference purposes.

**Comment by fallaha56** [comment]
hi, you're of course correct about reference renderers, perhaps have a software VI option as well (=reference)

re: semi-software: i got the impression from your blog that the RDP and VI are very separate, data only flows RDP->VI and that the VI is very do-able in OpenGL (if not done?)

plus there's performance: when disabling the VI in HatCat's Angrylion plugin i get a ~10fps boost

so, potential breakdown:
most accurate: LLE software RDP + software VI
accurate but accelerated: LLE software RDP + OpenGL VI
faster: LLE OpenGL GFX + VI
fastest: HLE

**Comment by purplemarshmallow** [comment]
@fallaha56 It's not hard to use angrylion's plugin with GLideN64's VI code. Here you go. Runs a bit faster than hatcat's plugin but there's not much difference.
https://drive.google.com/file/d/0B7Y6r4SpC_QQZ1pFMFJRdXFKbDQ/view?usp=sharing

@gonetz I combined GLideN64 and angrylion's plugin but I think it's only useful for debugging and comparing results. It's too slow and it's difficult to completely separate tasks in angrylion's plugin. If N64 depth compare is emulated on the GPU then the necessary data is available anyway. Just need to copy it over if needed

**Comment by purplemarshmallow** [comment]
>For me, some games are so bad on z64gl and Jabo's that I'm better off using a software plugin. It makes me question what is going on. I really should look into profiling gfx API code.

@LegendOfDragoon  Which games? For me z64gl is many times faster than angrylion's plugin

**Comment by fallaha56** [comment]
@purplemarshmallow thanks mate that's exactly what i was wondering -but doesn't look like the VI filters are being applied... ;((

is there anything i should be doing to turn them on? or should they just work?

![hatcatvion](https://cloud.githubusercontent.com/assets/11017537/12221593/f8c4560e-b799-11e5-94f5-4dcc784806b0.png)
![newbuild](https://cloud.githubusercontent.com/assets/11017537/12221592/f8be4296-b799-11e5-924a-757cec42869c.png)

**Comment by purplemarshmallow** [comment]
There are no VI filters they are not emulated. GLideN64's VI does not support that. But gamma correction is emulated I need to test that with the recent gamma correction fixes.

**Comment by fallaha56** [comment]
LOL me -thanks for clarifying...

@gonetz is that something worthwhile adding/quick to add? appreciate it's prob time i donate again ;)

but do look at the VI/s: pretty much 100% speed; and that's an RSP interpreter on a 5 y/o CPU ;))

**Comment by LuigiBlood** [comment]
I got this plugin just to hope I can switch between HLE and LLE easily during emulation but nope. Can't do that.
Also, Talent Studio LLE, when I try to access the Talent menu (the red one), the video fucks up really bad unlike angrylion plugin. :D

**Comment by purplemarshmallow** [comment]
>I got this plugin just to hope I can switch between HLE and LLE easily during emulation but nope. Can't do that.

The problem is on PJ64's side the plugin can do that. If you switch from PJ64's RSP to cxd4's interpreter it works for me.

>Also, Talent Studio LLE, when I try to access the Talent menu (the red one), the video fucks up really bad unlike angrylion plugin. :D 

Sounds like a VI bug

**Comment by gonetz** [comment]
Experimental branch with Glide64 soft depth buffer render used in LLE mode:
https://github.com/gonetz/GLideN64/tree/sw_depth_render
Initial implementation. Not tested it much.

**Comment by purplemarshmallow** [comment]
Just to remember software depth buffer rendering in Glide64 is buggy part of the depth image is not rendered see #677 implementing it will likely bring bugs over to GLideN64.

Maybe it's worth to give integer math in GLideN64's depth shader another try.

**Comment by LegendOfDragoon** [comment]
>@LegendOfDragoon Which games? For me z64gl is many times faster than angrylion's plugin

Vigilante 8 is like 2 VI/s with z64gl for me. Angrylions is at least 10 VI /s. To clarify, I'm mostly refering to when filters are disabled. There are other games where the speed difference is so little that I might as well use Angrylion's for the extra accuracy. Games such as SD Hiryuu, Dark Rift, and Last Legion come to mind. In fact, SD Hiryuu has really slow menus in z64gl and it crashes on your fork, unless i did something wrong. Angrylion's plugin is easy to optimize, so that's another reason why I said I'm better off using Angrylion's for some games. It's the most suitable plugin for me in SD Hiryuu, Last Legion, Dark Rift, etc. I know other people get better speeds with z64gl than I do, but I still am convinced there is something wrong with the code.

>but do look at the VI/s: pretty much 100% speed; and that's an RSP interpreter on a 5 y/o CPU ;))

Your post proved my point about popularity. You're not using the fastest RSP.

>re: semi-software: i got the impression from your blog that the RDP and VI are very separate, data only flows RDP->VI and that the VI is very do-able in OpenGL (if not done?)

It is not very doable. Coverage will not be easy to implement, if even feasible at all.

>The problem is on PJ64's side the plugin can do that. If you switch from PJ64's RSP to cxd4's interpreter it works for me.

Last time I checked, some games had glitches when switching between HLE and LLE, like Turok.

>Maybe it's worth to give integer math in GLideN64's depth shader another try.

I agree.

**Comment by purplemarshmallow** [comment]
>SD Hiryuu has really slow menus in z64gl and it crashes on your fork, unless i did something wrong.

I need to check is it caused by my changes or by the compiler switch.

>I know other people get better speeds with z64gl than I do, but I still am convinced there is something wrong with the code.

There is certainly something wrong with the code. If TMEM emulation is incorrect it can slow things down. See F-Zero X

>It is not very doable. Coverage will not be easy to implement, if even feasible at all.

What does pixel coverage have to do with the VI? I plugged angrylion's plugin into GLideN64's VI and it works

**Comment by theboy181** [comment]
Is there a bin for GLideN64 with Angrylion implemented VI code available?  I would like to test it

**Comment by purplemarshmallow** [comment]
angrylion's plugin with GLideN64's VI was hardly any work I had the code from my experiments with mixing software and hardware rendering. GLideN64 with angrylion's VI should be possible. Would be nice to have. It could be used by default for GLideN64's cfb mode.

**Comment by LegendOfDragoon** [comment]
>What does pixel coverage have to do with the VI? I plugged angrylion's plugin into GLideN64's VI and it works

VI filters use coverage, iirc. I think it's better to optimize the VI filter code, before considering using OpenGL to do the filters..

>There is certainly something wrong with the code. If TMEM emulation is incorrect it can slow things down. See F-Zero X

Ya, nice job fixing the crash in F-ZERO.

**Comment by purplemarshmallow** [comment]
>VI filters use coverage, iirc. I think it's better to optimize the VI filter code, before considering using OpenGL to do the filters..

Yes that makes sense. Would still be useful to port angrylion's VI code to GLideN64 and skip the filters. In can be useful in cfb mode and maybe it can also help with bugs in the current VI code

>Ya, nice job fixing the crash in F-ZERO.

Thanks. There are still glitches in the menus unfortunately because the load_block command is still not correct.

BTW there is a benchmarking tool included with z64gl: z64bench. But I have not looked at it yet. z64gl also has a graphical debugger. Ziggy wrote it's linux only but should be easy to port to windows.

**Comment by LegendOfDragoon** [comment]
>Would still be useful to port angrylion's VI code to GLideN64 and skip the filters. In can be useful in cfb mode and maybe it can also help with bugs in the current VI code

I agree.

>Thanks. There are still glitches in the menus unfortunately because the load_block command is still not correct.

You're sure it's the load_block command? If I can narrow down which functions are causing X bugs, I'd be more inclined to look into it. I took a look at z64 and z64gl and the code is much different than Angrylion's. I came to the conclusion that I might as well write a new plugin from scratch than to fix a codebase with so many issues. I'd rather focus on optimizing accurate code than fixing inaccurate code, since the results I get are more guaranteed. At the same time, I'd love to be able to use z64gl for games where it's actually fast for me. So it really depends how much work it will take to fix some of the key issues. I don't expect to fix the triangle issue anytime soon, but I'd at least like to see some of those bugs inherited from z64 fixed.

>BTW there is a benchmarking tool included with z64gl: z64bench. But I have not looked at it yet. z64gl also has a graphical debugger. Ziggy wrote it's linux only but should be easy to port to windows.

Sounds interesting. Maybe I'll check out z64bench. The debugger will be nice to use as well.

**Comment by purplemarshmallow** [comment]
>You're sure it's the load_block command?

I'm 100% sure. Because I replaced the load_block command with MAME's implementation and it fixed this bug but broke textures ingame

>I came to the conclusion that I might as well write a new plugin from scratch than to fix a codebase with so many issues.

Partial rewrites are also possible. For example rewrite just the combiner or TMEM functions. Ziggy spent some time fighting driver issues and optimizing everything this work also must be done again in a new plugin.

**Comment by purplemarshmallow** [comment]
Also I'd look at the current MAME code it's more similar to z64

**Comment by LegendOfDragoon** [comment]
>I'm 100% sure. Because I replaced the load_block command with MAME's implementation and it fixed this bug but broke textures ingame

Oh, ok. Guess I can start with that, the next time i feel like taking a look.

>Partial rewrites are also possible. For example rewrite just the combiner or TMEM functions. Ziggy spent some time fighting driver issues and optimizing everything this work also must be done again in a new plugin.

That is true. I still feel that it may take too long to track down these issues. I wish others were on board to try and track down certain bugs.

>Also I'd look at the current MAME code it's more similar to z64

Good idea actually. Maybe seeing their commit history will be a good idea too, in case they have notes on particular improvements. I'll have to look around i guess.

**Comment by purplemarshmallow** [comment]
here is MAME's commit history: http://git.redump.net/mame/log/src/mame/video/n64.c
It might also help with improving the triangle command.

**Comment by LegendOfDragoon** [comment]
Thanks, I'll try checking it out this week. I think the biggest issue with the triangle command is that it's in floating point. I have no idea how to convert float to fixed and vice versa.

**Comment by fallaha56** [comment]
@LegendOfDragoon is any of this any help?

https://dolphin-emu.org/blog/2014/03/15/pixel-processing-problems/

**Comment by LegendOfDragoon** [comment]
>@LegendOfDragoon is any of this any help?

I've seen it already. I understand integer math and Angrylion's code very well. I do not know z64gl's code or OpenGL too well. That is the real problem for me.

If z64gl is computing the triangle commands in hardware, then I think you can potentially sacrifice some speed by doing the integer math on the cpu and perhaps convert the end results to floating point. If it's already doing all the triangle command code on cpu, then that should be easier to fix.

**Comment by purplemarshmallow** [comment]
z64gl's triangle command is very similar to the incorrect triangle command in the z64 software plugin. z64 can't draw lines properly and z64gl also can't draw lines properly.

The triangle command was modified to get upscaled triangles suitable for hardware rendering. But this is calculated on the CPU.

**Comment by fallaha56** [comment]
@LegendOfDragoon thanks for the insights -out of curiosity is the N64 doing integer calculations like the Gamecube? Why the need to convert between FP and int if it makes it complicated?

**Comment by gonetz** [comment]
>If z64gl is computing the triangle commands in hardware, then I think you can potentially sacrifice some speed by doing the integer math on the cpu and perhaps convert the end results to floating point.

It is how it currently works. All calculations are done with integer (actually fixed point) math, and then  the result converted to floats before sending to GL. The problem is somewhere in these fixed point calculations.

>Why the need to convert between FP and int if it makes it complicated?

floats have native support by both CPU and GPU. We can use fixed point arithmetic on CPU and on modern GPU, but fixed point precision used by N64 and GPU is generally different, so conversion will be necessary anyway.

**Comment by gonetz** [comment]
> Just to remember software depth buffer rendering in Glide64 is buggy part of the depth image is not rendered see #677 implementing it will likely bring bugs over to GLideN64.

Do you mean this issue?
https://github.com/gonetz/GLideN64/issues/677#issuecomment-136014246
Could you check it with sw_depth_render branch? I think that that problem should not appear in LLE.

**Comment by purplemarshmallow** [comment]
Yes I mean this issue. I can't get the sw_depth_render branch working. No depth buffer is rendered for me.

**Comment by gonetz** [comment]
Depth buffer copy option  should be enabled for depth rendering.

**Comment by purplemarshmallow** [comment]
I tried again and got it working but it's buggy
![1](https://cloud.githubusercontent.com/assets/7283660/12274345/c5c90aec-b96a-11e5-9506-930aaa0704be.jpg)

![2](https://cloud.githubusercontent.com/assets/7283660/12274348/c8c0efc6-b96a-11e5-86c9-9da88f6eacf7.jpg)

**Comment by gonetz** [comment]
Thanks! The task is not so obvious as I thought.

**Comment by LegendOfDragoon** [comment]
>For efficient copy color you need glBlitFramebuffer.

Actually at least on my hardware (Intel), z64gl has the fastest implementation of fb_read (when compared to any plugin that uses sync mode). I think glReadPixels is good enough.

**Comment by fallaha56** [comment]
Regarding Intel GFX, is/would this be helpful...some form of direct GPU memory access?

https://software.intel.com/sites/default/files/managed/86/b9/Tiling_Format_for_Intel_Iris_Graphics_Extension_for_Instant_Access.pdf

**Comment by LegendOfDragoon** [comment]
I briefly read that page, and it seems to be for iris only. I don't think most Intel users have iris graphics and I certainly do not. It does seem useful though.

**Comment by purplemarshmallow** [comment]
>>For efficient copy color you need glBlitFramebuffer.

>Actually at least on my hardware (Intel), z64gl has the fastest implementation of fb_read (when compared to any plugin that uses sync mode). I think glReadPixels is good enough.

The trick is to scale the image down to N64 resolution before reading (or render in native res). GLideN64 uses glBlitFramebuffer to scale the the image down and then glReadPixels to read it from the GPU

z64gl uses a copy shader to draw a smaller version of the texture on screen and then it reads it. The problem is it can mess up the display if you use fb read at the wrong time because stuff is drawn on screen

**Comment by loganmc10** [comment]
>GLideN64 uses glBlitFramebuffer to scale the the image down and then glReadPixels to read it from the GPU
> z64gl uses a copy shader to draw a smaller version of the texture on screen and then it reads it.

I believe GLideN64 does the same thing for GLES2 since glBlitFramebuffer isn't available:

https://github.com/gonetz/GLideN64/blob/485f2307efbe4ff545cefcc55d312737fdd4608b/src/OpenGL.cpp#L2253-L2345

**Comment by LegendOfDragoon** [comment]
>The trick is to scale the image down to N64 resolution before reading (or render in native res). 

Also wouldn't converting the format to 16 bit also help, for 16bit color games?

>z64gl uses a copy shader to draw a smaller version of the texture on screen and then it reads it.

Interesting. I somehow thought glReadPixels supports scaling.

>The problem is it can mess up the display if you use fb read at the wrong time because stuff is drawn on screen

Wouldn't that apply to any plugin? What games is this an issue for? So far, I've tested SSB, Mario Kart, 1080 Snowboarding, and Banjo Kazooie and it seems to work fine for those 4.

**Comment by purplemarshmallow** [comment]
>I believe GLideN64 does the same thing for GLES2 since glBlitFramebuffer isn't available:
@loganmc10 yes this code seems to work very similar

>Also wouldn't converting the format to 16 bit also help, for 16bit color games?

I'm not sure if it's possible to convert a texture to 16 bit on the gpu

>Wouldn't that apply to any plugin?

It certainly applies to Glide64 it must draw everything on screen to read to. GLideN64 on Desktop does not need to draw anything because of glBlitFramebuffer so there's no problem. Maybe I can draw things offscreen with fbos, I need to test...

>What games is this an issue for? So far, I've tested SSB, Mario Kart, 1080 Snowboarding, and Banjo Kazooie and it seems to work fine for those 4.

Not sure if it affects these games. I wanted to test adding accurate VI emulation to z64gl (copy everything to RDRAM and then the VI reads from RDRAM) but the framebuffer2rdram function messed up everthing

**Comment by LegendOfDragoon** [comment]
I tested banjo some more on 1964 with z64gl. I got it to mess up on the puzzle pieces once, so I guess there's work to be done on the fb_read implementation. Not entirely sure on whether or not it's related to the issue you described though. I believe it has trouble filtering the extra fb_read calls generated from 1964. It also crashes in Banjo on mupen64 0.5.

I guess the scaling method used, is why z64gl's fb read is relatively fast.

>I'm not sure if it's possible to convert a texture to 16 bit on the gpu

Should be possible, I think.

>It certainly applies to Glide64 it must draw everything on screen to read to. GLideN64 on Desktop does not need to draw anything because of glBlitFramebuffer so there's no problem. Maybe I can draw things offscreen with fbos, I need to test...

I see. It's starting to make more sense to me now.

>I wanted to test adding accurate VI emulation to z64gl (copy everything to RDRAM and then the VI reads from RDRAM) but the framebuffer2rdram function messed up everthing

does your cfb code work? I might give it a try, soon. I recently took a brief look at z64gl's code and it seems quite complex.

**Comment by purplemarshmallow** [comment]
My cfb code works

Yes I think it should be possible to read from a RGBA_5_5_5_1 16 bit texture to speed up copy color for 16 bit games

**Comment by LegendOfDragoon** [comment]
I ended up trying out format conversion to 16bit with D3D9 and saw a nice perf boost, for FB read on my desktop. It didn't seem to help as much on laptop though. I'd say it's worth trying out. Since most games use 16bit framebuffer this should come in handy.
